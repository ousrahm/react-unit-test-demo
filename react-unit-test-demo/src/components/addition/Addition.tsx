import React from "react";
import styles from "./Addition.module.css";
import { Col, Container, Row } from "react-bootstrap";

const Addition = (): JSX.Element => {
  const [sum, setSum] = React.useState(0);
  const [addend1, setAddend1] = React.useState(0);
  const [addend2, setAddend2] = React.useState(0);

  React.useEffect(() => {
    setSum(addend1 + addend2);
  }, [addend1, addend2]);

  return (
    <div>
      <h2 className={styles.heading}>Addition Component</h2>
      <Container>
        <Row>
          <Col className={styles.column}>
            <input
              id="addend1"
              className={styles.input}
              type="number"
              defaultValue={addend1.toString()}
              onChange={(e) => setAddend1(parseInt(e.target.value, 10))}
            />
          </Col>
          <Col className={styles.column}>
            <span className={styles.text}>+</span>
          </Col>
          <Col className={styles.column}>
            <input
              id="addend2"
              className={styles.input}
              type="number"
              defaultValue={addend2.toString()}
              onChange={(e) => setAddend2(parseInt(e.target.value, 10))}
            />
          </Col>
          <Col className={styles.column}>
            <span className={styles.text}>=</span>
          </Col>
          <Col className={styles.column}>
            {isNaN(sum) ? (
              <span className={styles.warning}>
                Please enter a value in both inputs.
              </span>
            ) : (
              <span className={styles.sum}>{sum}</span>
            )}
          </Col>
        </Row>
      </Container>
    </div>
  );
};

export default Addition;
