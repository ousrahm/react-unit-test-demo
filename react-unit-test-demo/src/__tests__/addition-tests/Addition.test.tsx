import {
  act,
  fireEvent,
  render,
  screen,
  waitFor,
} from "@testing-library/react";
import Addition from "../../components/addition/Addition";

describe("Addition component tests", () => {
  it("renders without crashing", () => {
    render(<Addition />);
  });

  it("matches snapshot", () => {
    const addition = render(<Addition />).baseElement;

    expect(addition).toMatchSnapshot();
  });

  it("displays warning if one or both inputs are empty and non-zero", async () => {
    render(<Addition />);

    const addendInputs = await screen.findAllByDisplayValue("0");

    addendInputs.forEach((a) => fireEvent.change(a, { target: { value: "" } }));

    const warningMessage = await screen.findByText(
      "Please enter a value in both inputs."
    );
    expect(warningMessage).toBeInTheDocument();
  });

  it("displays correct sum with non-zero non-empty inputs", async () => {
    render(<Addition />);

    const addendInputs = await screen.findAllByDisplayValue("0");

    addendInputs.forEach((a) =>
      fireEvent.change(a, { target: { value: "5" } })
    );

    const sum = await screen.findByText("10");
    expect(sum).toBeInTheDocument();
  });
});
