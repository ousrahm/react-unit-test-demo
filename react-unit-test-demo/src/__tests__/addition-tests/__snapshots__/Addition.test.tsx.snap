// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Addition component tests matches snapshot 1`] = `
<body>
  <div>
    <div>
      <h2
        class="heading"
      >
        Addition Component
      </h2>
      <div
        class="container"
      >
        <div
          class="row"
        >
          <div
            class="column col"
          >
            <input
              class="input"
              id="addend1"
              type="number"
              value="0"
            />
          </div>
          <div
            class="column col"
          >
            <span
              class="text"
            >
              +
            </span>
          </div>
          <div
            class="column col"
          >
            <input
              class="input"
              id="addend2"
              type="number"
              value="0"
            />
          </div>
          <div
            class="column col"
          >
            <span
              class="text"
            >
              =
            </span>
          </div>
          <div
            class="column col"
          >
            <span
              class="sum"
            >
              0
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
